internal/application.go

```diff
@@ -0,0 +1,361 @@
+			currK := k
+			// run defer via anonymous function to not directly evaluate the arguments.
+			defer func() {
+				encoding.ReleaseKVEntry(currK)
+			}()
+
+			res.Headers = currK.ValueBytes()
+			// acquire a new kv entry to continue reading other message values.
+			k = encoding.AcquireKVEntry()
+		case "body":
+			// make a copy of the pointer and add a defer in case there is another entry
+			currK := k
+			// run defer via anonymous function to not directly evaluate the arguments.
+			defer func() {
+				encoding.ReleaseKVEntry(currK)
+			}()
+
+			res.Body = currK.ValueBytes()
+			// acquire a new kv entry to continue reading other message values.
+			k = encoding.AcquireKVEntry()
+		default:
+			a.Logger.Debug().Str("name", name).Msg("unknown kv entry")
+		}
+	}
+
+	if res.ID == "" {
+		return fmt.Errorf("response id is empty")
+	}
+
+	cv, ok := a.cache.Get(res.ID)
+	if !ok {
+		return fmt.Errorf("transaction %q not found", res.ID)
+	}
+	// TODO does remove forces eviction?
+	defer a.cache.Remove(res.ID)
+
+	tx := cv.(types.Transaction)
+
+	if err := readHeaders(res.Headers, tx.AddResponseHeader); err != nil {
+		return fmt.Errorf("reading headers: %v", err)
+	}
+
+	if it := tx.ProcessResponseHeaders(int(res.Status), "HTTP/"+res.Version); it != nil {
+		return ErrInterrupted{it}
+	}
+
+	switch it, _, err := tx.WriteResponseBody(res.Body); {
+	case err != nil:
+		return err
+	case it != nil:
+		return ErrInterrupted{it}
+	}
+
+	switch it, err := tx.ProcessResponseBody(); {
+	case err != nil:
+		return err
+	case it != nil:
+		return ErrInterrupted{it}
+	}
+
+	tx.ProcessLogging()
+	return tx.Close()
+}
+
+func (a AppConfig) NewApplication() (*Application, error) {
+	app := Application{
+		AppConfig: a,
+	}
+
+	config := coraza.NewWAFConfig().
+		WithDirectives(a.Directives).
+		WithErrorCallback(app.logCallback).
+		WithRootFS(mergefs.Merge(coreruleset.FS, io.OSFS))
+
+	waf, err := coraza.NewWAF(config)
+	if err != nil {
+		return nil, err
+	}
+	app.waf = waf
+
+	const defaultExpire = time.Second * 10
+	const defaultEvictionInterval = time.Second * 1
+
+	app.cache = cache.NewTTLWithCallback(defaultExpire, defaultEvictionInterval, func(key, value any) {
+		// everytime a transaction runs into a timeout it gets closed.
+		tx, ok := value.(types.Transaction)
+		if !ok {
+			return
+		}
+
+		// Process Logging won't do anything if TX was already logged.
+		tx.ProcessLogging()
+		if err := tx.Close(); err != nil {
+			a.Logger.Error().Err(err).Str("tx", tx.ID()).Msg("error closing transaction")
+		}
+	})
+
+	return &app, nil
+}
+
+func (a *Application) logCallback(mr types.MatchedRule) {
+	var l *zerolog.Event
+
+	switch mr.Rule().Severity() {
+	case types.RuleSeverityWarning:
+		l = a.Logger.Warn()
+	case types.RuleSeverityNotice,
+		types.RuleSeverityInfo:
+		l = a.Logger.Info()
+	case types.RuleSeverityDebug:
+		l = a.Logger.Debug()
+	default:
+		l = a.Logger.Error()
+	}
+	l.Msg(mr.ErrorLog())
```

Comment:
There might be a problem with the JSON log-format.
The spoa logs are in JSON, but the ones from coraza are not. (`SecAuditLogFormat JSON` is ignored)
It works with basic coraza 3.2.1
Some more info: https://github.com/corazawaf/coraza/issues/1150